p2p/src/supervisor.rs

```diff
@@ -0,0 +1,418 @@
+    /// later point then when the command was issued. Protocols which rely on hard upper bounds
+    /// like the number of concurrently connected peers should issue a disconnect to remedy the
+    /// situation.
+    Accept,
+    /// Establishes a new connection to the remote end in [`transport::ConnectInfo`].
+    Connect(transport::ConnectInfo),
+    /// Disconnects the [`peer::Peer`] known by [`node::Id`]. This will tear down the entire tree of
+    /// subroutines managing the peer in question.
+    Disconnect(node::Id),
+    /// Dispatch the given message to the peer known for [`node::Id`].
+    Msg(node::Id, message::Send),
+}
+
+/// Set of significant events in the p2p subsystem.
+pub enum Event {
+    /// A new connection has been established.
+    Connected(node::Id, Direction),
+    /// A [`peer::Peer`] has been disconnected.
+    Disconnected(node::Id, Report),
+    /// A new [`message::Receive`] from the [`peer::Peer`] has arrived.
+    Message(node::Id, message::Receive),
+    /// A connection upgraded successfully to a [`peer::Peer`].
+    Upgraded(node::Id),
+    /// An upgrade from failed.
+    UpgradeFailed(node::Id, Report),
+}
+
+enum Internal {
+    Accept,
+    Connect(transport::ConnectInfo),
+    SendMessage(node::Id, message::Send),
+    Stop(node::Id),
+    Upgrade(node::Id),
+}
+
+enum Input {
+    Accepted(node::Id),
+    Command(Command),
+    Connected(node::Id),
+    Receive(node::Id, message::Receive),
+    Stopped(node::Id, Option<Report>),
+    Upgraded(node::Id),
+    UpgradeFailed(node::Id, Report),
+}
+
+enum Output {
+    Event(Event),
+    Internal(Internal),
+}
+
+impl From<Event> for Output {
+    fn from(event: Event) -> Self {
+        Output::Event(event)
+    }
+}
+
+impl From<Internal> for Output {
+    fn from(internal: Internal) -> Self {
+        Output::Internal(internal)
+    }
+}
+
+/// Wrapping a [`transport::Transport`] the `Supervisor` runs the p2p machinery to manage peers over
+/// physical connections. Offering multiplexing of ingress and egress messages and a surface to
+/// empower higher-level protocols to control the behaviour of the p2p substack.
+pub struct Supervisor {
+    command_tx: Sender<Command>,
+    event_rx: Receiver<Event>,
+}
+
+struct State<Conn> {
+    connected: HashMap<node::Id, transport::Direction<Conn>>,
+    peers: HashMap<node::Id, peer::Peer<peer::Running<Conn>>>,
+}
+
+impl Supervisor {
+    /// Takes the [`transport::Transport`] and sets up managed subroutines. When the `Supervisor`
+    /// is returned the p2p subsystem has been successfully set up on the given network interface
+    /// (as far as applicable for the transport) so the caller can use the command input and
+    /// consume events.
+    ///
+    /// # Errors
+    ///
+    /// * If the bind of the transport fails
+    #[allow(clippy::too_many_lines)]
+    pub fn run<T>(transport: &T) -> Result<Self>
+    where
+        T: transport::Transport + Send + 'static,
```

Comment:
why `T` (transport) has a static lifetime?

Reply:
Could you provide more context about what makes this lifetime so surprising?

Reply:
I am not fully sure what happens there but it seems that, while `transport` itself is not explicitly moved between threads, some values of type `<T as Transport>::Endpoint` and `<T as Transport>::Connection` are. Rust apparently requires a `'static` bound on `T` for the types above to be considered `'static` as well. Not fully clear to me why that is.

On the other hand, the `Send` bound does not seem like it's needed (ie. one can remove it and the code still compiles).

Another way of working around that, would be to remove the `'static` bound here, and add it to the associated types of `Transport`:

```rust
pub trait Transport {
    type Connection: Connection + 'static;
    type Endpoint: Endpoint<Connection = Self::Connection> + 'static;
    type Incoming: Iterator<Item = Result<Self::Connection>> + Send + 'static;
```

but I am not sure which approach is more modular.

---

Just in case, some background for why the `'static` bound is needed for moving values of a generic types between threads: https://users.rust-lang.org/t/why-does-thread-spawn-need-static-lifetime-for-generic-bounds/4541/2

Reply:
Besides @romac alternative another version would be:

``` rust
    pub fn run<T>(transport: &T, info: transport::BindInfo) -> Result<Self>
    where
        T: transport::Transport + Send,
        T::Connection: 'static,
        T::Endpoint: 'static,
        T::Incoming: 'static,
```

as I generally prefer lifetime bounds to the latest possible moment. Yet at that point it doesn't really make a difference from constraining `Transport` directly.

> yeah, so from my newbie perspective `'static` lifetime should be applied sparingly. Does the transport really have to live for the duration of the entire program? I guess it's fine as only a few transports will be created, but still if it's possible to avoid `'static` I'd remove it.

Glad you bring that up as this is a common misconception. Personally I greatly improved my intuition of lifetimes by reading the sections in the Rustonomicon: https://doc.rust-lang.org/nomicon/lifetimes.html and this great piece: https://github.com/pretzelhammer/rust-blog/blob/master/posts/common-rust-lifetime-misconceptions.md

In this case we could change the lifetime to be named, but that would will trickle through the entire assortment of types and will need to be carried almost everywhere. satisfied.

Reply:
> yeah, so from my newbie perspective `'static` lifetime should be applied sparingly. Does the transport really have to live for the duration of the entire program?

That's not what a `'static` lifetime _bound_ means. There is a difference between a `'static` _lifetime_ and a `'static` _lifetime bound_. The former indeed means that the reference is valid for the entire duration of the program, while the latter means that the type bounded by it does not contain any reference which has a lifetime less than `'static`.

For example, the following types satisfy the `'static` lifetime bound:

```rust
u32
struct Foo(u32);
String
enum { Bar(String), Baz(bool) }
struct Test { x: &'static Foo }
```

On the other hand, this type will not satisfy the `'static` bound unless `'a == 'static`:

```rust
struct Abc<'a> {
  x: &'a i32
}
```

eg.

```rust
fn foo<'a>(x: Abc<'a>) {
  // Abc<'a> does not satisfy the 'static bound because we don't know anything about 'a (it's universally quantified)
}

fn bar(x: Abc<'static>) {
  // Abc<'static> satisfies the 'static bound because the reference it contains is 'static
}
````

Hope that clears things up :)

Reply:
Thanks @xla for the link! This section in particular explains this stuff in a much better way than I did above: https://github.com/pretzelhammer/rust-blog/blob/master/posts/common-rust-lifetime-misconceptions.md#2-if-t-static-then-t-must-be-valid-for-the-entire-program

Reply:
> Besides @romac alternative another version would be:
> 
> ```rust
>     pub fn run<T>(transport: &T, info: transport::BindInfo) -> Result<Self>
>     where
>         T: transport::Transport + Send,
>         T::Connection: 'static,
>         T::Endpoint: 'static,
>         T::Incoming: 'static,
> ```
> 
> as I generally prefer lifetime bounds to the latest possible moment. Yet at that point it doesn't really make a difference from constraining `Transport` directly.

Agreed, that sounds like the best solution to me.